### ISSUE TITLE
[Bug]: Continuous aggregate sequential scan causes poor performance

### ISSUE THREAD

**Issue by idoberko2** [timescale/timescaledb#3895]
### What type of bug is this?

Performance issue

### What subsystems and features are affected?

Continuous aggregate

### What happened?

I have a hypertable defined like so:
```sql
CREATE TABLE feature_values(
    timestamp       TIMESTAMPTZ       	NOT NULL,
    model_id		VARCHAR				NOT NULL,
    feature_key    	VARCHAR             NOT NULL,
    feature_type    	SMALLINT             NOT NULL,
    feature_num_val DOUBLE PRECISION    NULL,
    feature_str_val VARCHAR             NULL,
);
ALTER TABLE feature_values
    ADD CONSTRAINT feature_vals_feature_uniq PRIMARY KEY (timestamp, model_id, feature_key);
```

And a daily aggregate of the above hypertable:
```sql
CREATE MATERIALIZED VIEW daily_features WITH (timescaledb.continuous)
    AS SELECT
        model_id,
        feature_key,
        feature_type,
        feature_str_val,
        time_bucket('1 day', timestamp) AS ts,
    	MIN(feature_num_val) AS feature_min,
    	MAX(feature_num_val) AS feature_max,
    	percentile_agg(feature_num_val) AS percentile_agg,
        COUNT(feature_str_val) AS value_count,
        COUNT(feature_key) AS exist_count
    FROM feature_values
    GROUP BY model_id, feature_key, feature_type, ts, feature_str_val;
```

For some reason, queries started having poor performance after inserting a large amount of rows (~1M) into `feature_values` hypertable.

I imagined a simple select statement filtering by only the `timestamp` to use only index scans but for some reasons, many sequential scans are performed:

```sql
EXPLAIN ANALYZE
SELECT * FROM daily_features WHERE ts='2021-12-04 00:00:00+00';
```
result:
```
Append  (cost=0.00..557392.83 rows=6116 width=164) (actual time=12214.060..16017.430 rows=150 loops=1)
  ->  HashAggregate  (cost=0.00..0.02 rows=1 width=234) (actual time=1628.181..1628.184 rows=0 loops=1)
        Group Key: model_id, feature_key, feature_type, ts, feature_str_val
        Batches: 1  Memory Usage: 32kB
        ->  Result  (cost=0.00..0.00 rows=0 width=0) (actual time=1628.176..1628.177 rows=0 loops=1)
              One-Time Filter: false
  ->  GroupAggregate  (cost=555221.96..557301.06 rows=6115 width=164) (actual time=10585.875..14388.946 rows=150 loops=1)
"        Group Key: feature_values.model_id, feature_values.feature_key, feature_values.feature_type, (time_bucket('1 day'::interval, feature_values.""timestamp"")), feature_values.feature_str_val"
        ->  Sort  (cost=555221.96..555374.84 rows=61150 width=108) (actual time=10585.825..12935.630 rows=1470018 loops=1)
              Sort Key: feature_values.model_id, feature_values.feature_key, feature_values.feature_type, feature_values.feature_str_val
              Sort Method: external merge  Disk: 173688kB
              ->  Custom Scan (ChunkAppend) on feature_values  (cost=0.00..550360.52 rows=61150 width=108) (actual time=950.836..4756.717 rows=1470018 loops=1)
                    Chunks excluded during startup: 0
                    ->  Seq Scan on _hyper_2_2_chunk  (cost=0.00..17048.81 rows=1887 width=108) (actual time=52.362..52.363 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 377327
                    ->  Seq Scan on _hyper_2_6_chunk  (cost=0.00..29948.60 rows=3354 width=104) (actual time=92.859..92.860 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 670726
                    ->  Seq Scan on _hyper_2_10_chunk  (cost=0.00..1107.70 rows=127 width=96) (actual time=4.289..4.290 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 25409
                    ->  Seq Scan on _hyper_2_14_chunk  (cost=0.00..2777.11 rows=313 width=102) (actual time=7.474..7.475 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 62538
                    ->  Seq Scan on _hyper_2_18_chunk  (cost=0.00..30053.80 rows=3389 width=101) (actual time=83.483..83.484 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 677769
                    ->  Seq Scan on _hyper_2_22_chunk  (cost=0.00..73590.97 rows=8081 width=112) (actual time=206.001..206.002 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 1616309
                    ->  Seq Scan on _hyper_2_26_chunk  (cost=0.00..1.07 rows=1 width=178) (actual time=0.010..0.011 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                    ->  Seq Scan on _hyper_2_29_chunk  (cost=0.00..1.41 rows=1 width=178) (actual time=0.005..0.005 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                    ->  Seq Scan on _hyper_2_37_chunk  (cost=0.00..2850.09 rows=327 width=96) (actual time=8.662..8.663 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 65426
                    ->  Seq Scan on _hyper_2_41_chunk  (cost=0.00..9660.60 rows=1109 width=96) (actual time=25.827..25.828 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221760
                    ->  Seq Scan on _hyper_2_45_chunk  (cost=0.00..9660.60 rows=1109 width=96) (actual time=26.654..26.654 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221760
                    ->  Seq Scan on _hyper_2_49_chunk  (cost=0.00..9660.60 rows=1109 width=96) (actual time=26.359..26.360 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221760
                    ->  Seq Scan on _hyper_2_53_chunk  (cost=0.00..9660.60 rows=1109 width=96) (actual time=26.484..26.485 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221760
                    ->  Seq Scan on _hyper_2_57_chunk  (cost=0.00..9660.65 rows=1109 width=96) (actual time=25.392..25.393 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221762
                    ->  Seq Scan on _hyper_2_61_chunk  (cost=0.00..9660.56 rows=1109 width=96) (actual time=24.262..24.263 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221758
                    ->  Seq Scan on _hyper_2_64_chunk  (cost=0.00..9660.65 rows=1109 width=96) (actual time=25.516..25.516 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221762
                    ->  Seq Scan on _hyper_2_68_chunk  (cost=0.00..9660.60 rows=1109 width=96) (actual time=28.841..28.842 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221760
                    ->  Seq Scan on _hyper_2_72_chunk  (cost=0.00..9660.56 rows=1109 width=96) (actual time=27.318..27.318 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221758
                    ->  Seq Scan on _hyper_2_75_chunk  (cost=0.00..6043.51 rows=694 width=96) (actual time=22.496..22.498 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 138734
                    ->  Seq Scan on _hyper_2_81_chunk  (cost=0.00..9684.88 rows=1109 width=96) (actual time=24.747..24.748 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221817
                    ->  Seq Scan on _hyper_2_88_chunk  (cost=0.00..1.07 rows=1 width=178) (actual time=0.032..0.033 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 3
                    ->  Seq Scan on _hyper_2_92_chunk  (cost=0.00..1.18 rows=1 width=178) (actual time=0.015..0.017 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 8
                    ->  Seq Scan on _hyper_2_96_chunk  (cost=0.00..3161.82 rows=361 width=96) (actual time=10.869..10.870 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 72170
                    ->  Seq Scan on _hyper_2_102_chunk  (cost=0.00..21.25 rows=2 width=96) (actual time=0.006..0.006 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                    ->  Seq Scan on _hyper_2_107_chunk  (cost=0.00..275775.18 rows=30240 width=113) (actual time=200.815..1906.020 rows=1470018 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 4578213
                    ->  Seq Scan on _hyper_2_112_chunk  (cost=0.00..1501.20 rows=172 width=96) (actual time=3.947..3.948 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 34320
                    ->  Seq Scan on _hyper_2_116_chunk  (cost=0.00..9692.60 rows=1109 width=96) (actual time=26.064..26.065 rows=0 loops=1)
"                          Filter: ((time_bucket('1 day'::interval, ""timestamp"") = '2021-12-04 00:00:00+00'::timestamp with time zone) AND (""timestamp"" >= COALESCE(_timescaledb_internal.to_timestamp(_timescaledb_internal.cagg_watermark(48)), '-infinity'::timestamp with time zone)))"
                          Rows Removed by Filter: 221760
Planning Time: 2.950 ms
JIT:
  Functions: 122
  Options: Inlining true, Optimization true, Expressions true, Deforming true
  Timing: Generation 16.429 ms, Inlining 16.160 ms, Optimization 978.152 ms, Emission 634.307 ms, Total 1645.048 ms
Execution Time: 16062.914 ms
```

### TimescaleDB version affected

2.5.0

### PostgreSQL version used

13.4

### What operating system did you use?

docker image timescale/timescaledb-ha:pg13.4-ts2.5.0-p0

### What installation method did you use?

Docker

### What platform did you run on?

Amazon Web Services (AWS)

### Relevant log output and stack trace

```bash
Don't have logs
```

### How can we reproduce the bug?

```bash
Follow the steps I provided above
```

**Comment by idoberko2** [comment]
I believe this is not an issue but a lack of understanding of CAGG functionality + real time aggregates on my side. The fundamental issue was I did not set up a refresh policy on my CAGG. Therefore, all queries to that CAGG effectively scanned the raw data instead of the non-existing materialized data.

I saw an immediate performance improvement once setting up the refresh policy and this is no longer an issue for me.

I think it might be worth to stress out in the docs more that CAGG doesn't actually do what you think it does unless you set up a refresh policy, or maybe even changing the API so that it forces the user to set a refresh policy (don't know if that's an option), for example:

```sql
CREATE MATERIALIZED VIEW hourly_data WITH (timescaledb.continuous)
    AS SELECT
        ...
    FROM raw_data
    GROUP BY hour
    WITH refresh_policy(...);
```
